File: examples/companion-packages/cpp-conan-app/src/variants.cpp
#include "pawmatch/variants.hpp"

#include <stdexcept>

namespace pawmatch {

MatchStrategy parse_match_strategy(const std::string& value) {
    if (value == "popularity") return MatchStrategy::Popularity;
    if (value == "match-quiz") return MatchStrategy::MatchQuiz;
    if (value == "longest-stay") return MatchStrategy::LongestStay;
    throw std::invalid_argument("unknown recommendation-strategy: " + value);
}

MatchDepth parse_match_depth(const std::string& value) {
    if (value == "short") return MatchDepth::Short;
    if (value == "standard") return MatchDepth::Standard;
    if (value == "thorough") return MatchDepth::Thorough;
    throw std::invalid_argument("unknown match-quiz-depth: " + value);
}

PetCardStyle parse_pet_card_style(const std::string& value) {
    if (value == "compact") return PetCardStyle::Compact;
    if (value == "detailed") return PetCardStyle::Detailed;
    if (value == "playful") return PetCardStyle::Playful;
    throw std::invalid_argument("unknown pet-card-style: " + value);
}

DonateFocus parse_donate_focus(const std::string& value) {
    if (value == "all") return DonateFocus::All;
    if (value == "shelters") return DonateFocus::Shelters;
    if (value == "rescue") return DonateFocus::Rescue;
    if (value == "policy") return DonateFocus::Policy;
    throw std::invalid_argument("unknown donate-focus-default: " + value);
}

const char* to_string(MatchStrategy v) {
    switch (v) {
        case MatchStrategy::Popularity: return "popularity";
        case MatchStrategy::MatchQuiz: return "match-quiz";
        case MatchStrategy::LongestStay: return "longest-stay";
    }
    return "match-quiz";
}

const char* to_string(MatchDepth v) {
    switch (v) {
        case MatchDepth::Short: return "short";
        case MatchDepth::Standard: return "standard";
        case MatchDepth::Thorough: return "thorough";
    }
    return "standard";
}

const char* to_string(PetCardStyle v) {
    switch (v) {
        case PetCardStyle::Compact: return "compact";
        case PetCardStyle::Detailed: return "detailed";
        case PetCardStyle::Playful: return "playful";
    }
    return "detailed";
}

const char* to_string(DonateFocus v) {
    switch (v) {
        case DonateFocus::All: return "all";
        case DonateFocus::Shelters: return "shelters";
        case DonateFocus::Rescue: return "rescue";
        case DonateFocus::Policy: return "policy";
    }
    return "all";
}

}  // namespace pawmatch

File: examples/companion-packages/golang-app/internal/pawmatch/variants.go
package pawmatch

import "fmt"

// Variant value types — narrowed strings so the CLI can switch on them.
type (
	// PetCardStyle is the rendering style for the browse listing.
	PetCardStyle string
	// MatchStrategy is the algorithm used to rank pets in `match`.
	MatchStrategy string
	// MatchDepth controls how many questionnaire factors `match` considers.
	MatchDepth string
	// DonateFocus is the default charity focus when `--focus` is not given.
	DonateFocus string
)

// Pet-card styles.
const (
	PetCardCompact  PetCardStyle = "compact"
	PetCardDetailed PetCardStyle = "detailed"
	PetCardPlayful  PetCardStyle = "playful"
)

// Match strategies.
const (
	StrategyPopularity  MatchStrategy = "popularity"
	StrategyMatchQuiz   MatchStrategy = "match-quiz"
	StrategyLongestStay MatchStrategy = "longest-stay"
)

// Match depths.
const (
	DepthShort    MatchDepth = "short"
	DepthStandard MatchDepth = "standard"
	DepthThorough MatchDepth = "thorough"
)

// Donate focuses.
const (
	FocusAll      DonateFocus = "all"
	FocusShelters DonateFocus = "shelters"
	FocusRescue   DonateFocus = "rescue"
)

var (
	petCardStyles = []PetCardStyle{PetCardCompact, PetCardDetailed, PetCardPlayful}
	strategies    = []MatchStrategy{StrategyPopularity, StrategyMatchQuiz, StrategyLongestStay}
	depths        = []MatchDepth{DepthShort, DepthStandard, DepthThorough}
	focuses       = []DonateFocus{FocusAll, FocusShelters, FocusRescue}
)

// ParsePetCardStyle decodes a TinyFlags variant value into the typed enum.
func ParsePetCardStyle(value string) (PetCardStyle, error) {
	for _, v := range petCardStyles {
		if string(v) == value {
			return v, nil
		}
	}
	return "", fmt.Errorf("pawmatch: unknown pet-card-style variant %q", value)
}

// ParseMatchStrategy decodes a TinyFlags variant value into the typed enum.
func ParseMatchStrategy(value string) (MatchStrategy, error) {
	for _, v := range strategies {
		if string(v) == value {
			return v, nil
		}
	}
	return "", fmt.Errorf("pawmatch: unknown recommendation-strategy variant %q", value)
}

// ParseMatchDepth decodes a TinyFlags variant value into the typed enum.
func ParseMatchDepth(value string) (MatchDepth, error) {
	for _, v := range depths {
		if string(v) == value {
			return v, nil
		}
	}
	return "", fmt.Errorf("pawmatch: unknown match-quiz-depth variant %q", value)
}

// ParseDonateFocus decodes a TinyFlags variant value into the typed enum.
func ParseDonateFocus(value string) (DonateFocus, error) {
	for _, v := range focuses {
		if string(v) == value {
			return v, nil
		}
	}
	return "", fmt.Errorf("pawmatch: unknown donate-focus-default variant %q", value)
}
